Classes/EventPublisher/JobQueueEventPublisher.php

```diff
@@ -0,0 +1,71 @@
+<?php
+declare(strict_types=1);
+namespace Neos\EventSourcing\EventPublisher;
+
+/*
+ * This file is part of the Neos.EventSourcing package.
+ *
+ * (c) Contributors of the Neos Project - www.neos.io
+ *
+ * This package is Open Source Software. For the full copyright and license
+ * information, please view the LICENSE file which was distributed with this
+ * source code.
+ */
+
+use Flowpack\JobQueue\Common\Job\JobManager;
+use Neos\EventSourcing\Event\DomainEvents;
+use Neos\EventSourcing\EventPublisher\JobQueue\CatchUpEventListenerJob;
+use Neos\EventSourcing\EventPublisher\Mapping\Mappings;
+use Neos\Flow\Annotations as Flow;
+
+/**
+ * An Event Publisher that sends events to a Job Queue using the Flowpack.JobQueue package.
+ *
+ * The queue Name
+ */
+final class JobQueueEventPublisher implements EventPublisherInterface
+{
+    private const DEFAULT_QUEUE_NAME = 'neos-eventsourcing';
+
+    /**
+     * @Flow\Inject
+     * @var JobManager
+     */
+    protected $jobManager;
+
+    /**
+     * @var string
+     */
+    private $eventStoreIdentifier;
+
+    /**
+     * @var Mappings
+     */
+    private $mappings;
+
+    public function __construct(string $eventStoreIdentifier, Mappings $mappings)
+    {
+        $this->eventStoreIdentifier = $eventStoreIdentifier;
+        $this->mappings = $mappings;
+    }
+
+    /**
+     * Enqueue domain events, such that their corresponding Event Listeners are executed lateron.
+     *
+     * @param DomainEvents $events
+     */
+    public function publish(DomainEvents $events): void
+    {
+        $queuedEventListenerClassNames = [];
+        foreach ($this->mappings->getMappingsForEvents($events) as $mapping) {
```

Comment:
I don't yet understand this code :)

Reply:
Mh, true.. Too magic. It's basically a filter that returns only mappings that are relevant for one of the given $events.

Do you think a nested foreach would be easier to understand?
```php
$queuedEventListenerClassNames = [];
$processedEventClassNames = [];
foreach ($events as $event) {
    $eventClassName = \get_class($event instanceof DecoratedEvent ? $event->getWrappedEvent() : $event);
    if (isset($processedEventClassNames[$eventClassName])) {
        continue;
    }
    $mappingsForEventClassName = $this->mappings->filter()...
    foreach ($mappingsForEventClassName as $mapping) {
            $listenerClassName = $mapping->getListenerClassName();
            if (isset($queuedEventListenerClassNames[$listenerClassName])) {
                continue;
            }
            $queueName = $mapping->getOption('queueName', self::DEFAULT_QUEUE_NAME);
            $options = $mapping->getOption('queueOptions', []);
            $this->jobManager->queue($queueName, new CatchUpEventListenerJob($listenerClassName, $this->eventStoreIdentifier), $options);
            $queuedEventListenerClassNames[$listenerClassName] = true;
    }
}
```

or do you have a different idea?

Reply:
OK, I adjusted the code now with 19942a3306fa52da2788427dac9ea79956dc5980 (and added a unit test to make sure that the end result is still the same).
It's still somewhat complex and maybe we can simplify this further in the future – but I would suggest to merge this one as is – unless you have a better idea on how to make that more readable of course!